### PR TITLE
gcc potential overflow and various lgtm scan fixes

### PR DESCRIPTION
**Pull request by kerolasa**
Both commits are overly pedantic, and done to keep static analysis noise level low so that real issues are easy to notice.

### CODE REVIEW

File: include/timeutils.h [hunk 1]
```diff
@@ -88,4 +90,13 @@ int strtime_short(const time_t *t, struct timeval *now, int flags, char *buf, si
...
 
+static inline char *xctime(time_t *t)
+{
+	char *buf;
+
+	buf = xcalloc(1, 64);
+	ctime_r(t, buf);
+	return buf;
+}
+
```
Comment (karelzak) [inline include/timeutils.h]: Not sure if I like it... would be better to use in code directly ctime_t and static buffer?

File: login-utils/login.c [hunk 2]
```diff
@@ -524,9 +525,13 @@ static void log_lastlog(struct login_context *cxt)
...
 		if (read(fd, (char *)&ll, sizeof(ll)) == sizeof(ll) &&
 							ll.ll_time != 0) {
+			char *time_string;
+
 			time_t ll_time = (time_t) ll.ll_time;
 
-			printf(_("Last login: %.*s "), 24 - 5, ctime(&ll_time));
+			time_string = xctime(&ll_time);
+			printf(_("Last login: %.*s "), 24 - 5, time_string);
+			free(time_string);
```
Comment (karelzak) [inline login-utils/login.c]:
```
char cbuf[26];
printf(_("Last login: %.*s "), 24 - 5, ctime_t(&ll_time, cbuf));
```
no malloc() and free() call at all.

File: login-utils/last.c [hunk 3]
```diff
@@ -339,15 +344,21 @@ static int time_formatter(int fmt, char *dst, size_t dlen, time_t *when)
...
 			ret = -1;
 		break;
 	}
 	case LAST_TIMEFTM_CTIME:
-		snprintf(dst, dlen, "%s", ctime(when));
+	{
+		char *s = xctime(when);
+		snprintf(dst, dlen, "%s", s);
+		free(s);
 		ret = rtrim_whitespace((unsigned char *) dst);
```
Comment (karelzak) [inline login-utils/last.c]:
```
char cbuf[26];
snprintf(dst, dlen, "%s", ctime(when, &cbuf));
```
no malloc() and free() call at all.

File: sys-utils/ipcs.c [hunk 4]
```diff
@@ -43,8 +43,14 @@ static void print_sem (int id);
 static void do_msg (char format, int unit);
 static void print_msg (int id, int unit);
 
-/* we read time as int64_t from /proc, so cast... */
-#define xctime(_x)	ctime((time_t *) (_x))
+static inline char *xctime(int64_t *t)
```
Comment (karelzak) [inline sys-utils/ipcs.c]: Maybe rename to ctime64().

**Comment by karelzak** [comment]
Sounds good, but xctime() seems like overkill.

**Comment by kerolasa** [comment]
> Sounds good, but xctime() seems like overkill.

Fair comment. The commit 035c89d3cb6fb8b9cb214f3d42af02faf1122da9 is allocating `char buf[26]` from stack when needed here and there in code.

**Pull request merged**